File: GluonTestResults/cpp/ITestClass.cpp
#include "PCH.h"
#include "ITestClass.h"

using namespace GluonTest;

ITestClass::ITestClass()
{
}

ITestClass::~ITestClass()
{
	_prop = 77;
}

int ITestClass::GetProperty() const
{
	return _prop;
}

void ITestClass::SetProperty(int value)
{
	_prop = value;
}

int ITestClass::GetReadOnlyProperty() const
{
	return 42;
}

int ITestClass::RetMethod()
{
    return DEFAULT_(int);
}

void ITestClass::Ref3(const TestStruct& thing)
{
}

Delegate<char(string)> ITestClass::GetDumbDelegate() const
{
    return DEFAULT_(Delegate<char(string)>);
}

void ITestClass::SetDumbDelegate(const Delegate<char(string)>& value)
{
}


AddSomeShit ITestClass::GetAdder() const
{
    return DEFAULT_(AddSomeShit);
}


void ITestClass::SetAdder(const AddSomeShit& value)
{
}


void ITestClass::Methody()
{
}


void ITestClass::OutMethod(int& outX)
{
}


void ITestClass::RefMethod(string& inoutThing)
{
}


Array<int> ITestClass::ComplexMethod(string& inoutA, IUnknown* _dumb, void* p, Array<char>& outFart)
{
    return DEFAULT_(Array<int>);
}


void ITestClass::Ref2(com_ptr<ITestClass>& inoutThing)
{
}


Array<TestStruct> ITestClass::GetHardProperty() const
{
    return DEFAULT_(Array<TestStruct>);
}


void ITestClass::SetHardProperty(Array<TestStruct> value)
{
}


Array<Delegate<string(char, int)>> ITestClass::GetHarderProperty() const
{
    return DEFAULT_(Array<Delegate<string(char, int)>>);
}


void ITestClass::SetHarderProperty(Array<Delegate<string(char, int)>> value)
{
}

File: GluonTest/ITestClass.cs
using Gluon;
using System;
using System.ComponentModel;
using System.Runtime.InteropServices;

namespace GluonTest
{
    // Supported constructs: 
    // Constructors
    // Methods
    // Properties
    // Read-Only Properties
    // Events
    // output parameters
    // ref parameters

    // Supported data types:
    // primitives
    // strings
    // blittable structs
    // Gluon interfaces/classes
    // IUnknown derivatives
    // arrays of any of these

    // Targets:
    // standard C#/.net
    // Unity would be cool (mono 2.x)
    // C++/Sharpish

    // Output:
    // .net library wrapping c++ library
    // C++ headers defining classes and interfaces
    // C++ skeleton source files

    // Need to be careful to segregate generated C++ from human C++

    public interface ITestClass
    {
        AddSomeShit Adder { get; set; }

        void Methody();
        int RetMethod();
        void OutMethod(out int x);
        int Property { get; set; }
        int ReadOnlyProperty { get; }
        TestStruct[] HardProperty { get; set; }
        Func<char,int,string>[] HarderProperty { get; set; }
        event Action<int> BigEvent;
        Func<string,char> DumbDelegate { get; set; }
        void RefMethod(ref string thing);
        void Ref2(ref ITestClass thing);
        void Ref3(TestStruct thing);
        int[] ComplexMethod(ref string a, IUnknown _dumb, IntPtr p, out char[] fart);
    }

    public enum Foo
    {
        Doo,
        Boo,
        Blu = 5,
        Noo = 3
    }

    [Native(header: "d3d11.h", staticLib: "d3d11.lib")]
    public interface ID3DBlob { }

    public struct TestStruct
    {
        public char a;
        public int b;
        public long c;
        public Int32 d;
        public string e;
        public int[] f;
    }

    public delegate ITestClass AddSomeShit(int a, int b);
}
